lpc55_sign/src/signed_image.rs

```diff
@@ -24,366 +22,123 @@ fn get_pad(val: usize) -> usize {
-
-    // the easiest way to get the bytes we need to sign is to read back
-    // what we just wrote
-    let sign_bytes = std::fs::read(outfile_path)?;
+    Ok(image_bytes)
+}
 
-    let mut img_hash = sha2::Sha256::new();
-    img_hash.update(&sign_bytes);
+/// Decode the private key, sign the stamped image with it,
+/// and append the signature to the image.
+pub fn sign_image(binary: &[u8], private_key: &str) -> Result<Vec<u8>> {
+    let mut image_hash = Sha256::new();
+    image_hash.update(binary);
 
-    let sig = priv_key.sign(
+    let private_key = RsaPrivateKey::from_pkcs1_pem(private_key)
+        .or_else(|_| RsaPrivateKey::from_pkcs8_pem(private_key))?;
+    let signature = private_key.sign(
         rsa::pkcs1v15::Pkcs1v15Sign::new::<rsa::sha2::Sha256>(),
-        img_hash.finalize().as_slice(),
+        image_hash.finalize().as_slice(),
     )?;
 
-    println!("Image signature {:x?}", sig);
-
-    let mut out = OpenOptions::new()
-        .write(true)
-        .append(true)
-        .open(outfile_path)?;
-
-    out.write_all(sig.as_ref())?;
-    drop(out);
-
-    Ok(rkth.as_slice().try_into().expect("something went wrong?"))
+    let mut signed = binary.to_owned();
+    signed.extend_from_slice(&signature);
+    Ok(signed)
 }
 
-pub fn create_cmpa(
-    with_dice: bool,
-    with_dice_inc_nxp_cfg: bool,
-    with_dice_cust_cfg: bool,
-    with_dice_inc_sec_epoch: bool,
-    rkth: &[u8; 32],
-    cmpa_path: &Path,
-) -> Result<()> {
-    let mut cmpa_out = OpenOptions::new()
-        .write(true)
-        .truncate(true)
-        .create(true)
-        .open(cmpa_path)?;
-
-    let mut secure_boot_cfg = SecureBootCfg::new();
-
-    secure_boot_cfg.set_dice(with_dice);
-    secure_boot_cfg.set_dice_inc_nxp_cfg(with_dice_inc_nxp_cfg);
-    secure_boot_cfg.set_dice_inc_cust_cfg(with_dice_cust_cfg);
-    secure_boot_cfg.set_dice_inc_sec_epoch(with_dice_inc_sec_epoch);
-    secure_boot_cfg.set_sec_boot(true);
-
-    let mut cmpa = CMPAPage::new();
-
-    cmpa.set_secure_boot_cfg(secure_boot_cfg)?;
-    cmpa.set_rotkh(rkth);
-
-    cmpa.set_debug_fields(DebugSettings::new())?;
-    cmpa.set_boot_cfg(DefaultIsp::Auto, BootSpeed::Fro96mhz)?;
-    let cmpa_bytes = cmpa.pack()?;
+pub fn root_key_hash(root: Vec<u8>) -> Result<[u8; 32]> {
+    if root.is_empty() {
+        Ok([0; 32])
+    } else {
+        let (_, root_cert) = parse_x509_certificate(&root)?;
+        let root_key = root_cert.public_key().subject_public_key.as_ref();
+        let root_key = RsaPublicKey::from_pkcs1_der(root_key)?;
+        let mut hash = Sha256::new();
+        hash.update(&root_key.n().to_bytes_be());
+        hash.update(&root_key.e().to_bytes_be());
+        Ok(hash.finalize().as_slice().try_into()?)
+    }
+}
 
-    cmpa_out.write_all(&cmpa_bytes)?;
-    Ok(())
+pub fn root_key_table_hash(root_certs: Vec<Vec<u8>>) -> Result<[u8; 32]> {
+    let mut rkth = Sha256::new();
+    for root in root_certs {
+        rkth.update(&root_key_hash(root)?);
+    }
```

Comment:
I don't think this is quite right: if we have fewer than 4 root certs, then we also need to hash `[0; 32]` (see [here](https://github.com/oxidecomputer/lpc55_support/blob/d0091d019dc568c369867446b453a5c35f0b0f5c/lpc55_sign/src/signed_image.rs#L183)) into the root hash.

Also, we should add a check that `root_certs.len() <= 4` here.

Reply:
Yeah, that should be fixed in b45706b with `pad_roots`.